doc/man/gnunet-download.1
.TH GNUNET-DOWNLOAD "1" "18 April Oct 2004" "GNUnet"
.SH NAME
gnunet-download \- a command line interface for downloading files from GNUnet
.SH SYNOPSIS
.B gnunet\-download
[\fIOPTIONS\fR] \-\- GNUNET_URI
.SH DESCRIPTION
.PP
Download files from GNUnet.
.TP
\fB\-a \fILEVEL\fR, \fB\-\-anonymity=LEVEL\fR
set additional receiver anonymity enforcing level. If unset,
value specified in config file is used. Default is 0.
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=FILENAME\fR
use config file (defaults: ~/.gnunet/gnunet.conf)
.TP
\fB\-o \fIFILENAME\fR, \fB\-\-output=FILENAME\fR
write the file to FILENAME (required)
.TP
\fB\-R\fR, \fB\-\-recursive\fR
download directories recursively and in parallel
.TP
\fB\-t\fINUMBER\fR, \fB\-\-threads=NUMBER\fR
when downloading directories recursively, this option specifies how many downloads are done in parallel.  The default is 30.  If you have not so much memory or if your system can not support about 64 threads, you may want to pass a smaller value here.  The default can be changed by adding an option PARALLELIZATION in the gnunet.conf file in the section [GNUNET-DOWNLOAD].
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number
.TP
\fB\-V\fR, \fB\-\-verbose\fR
print progress information
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=LOGLEVEL\fR
change the loglevel. Possible values for LOGLEVEL are NOTHING, FATAL, ERROR, FAILURE, WARNING, MESSAGE, INFO, DEBUG, CRON and EVERYTHING.
.SH NOTES
The GNUNET_URI is typically obtained from gnunet-search. gnunet-gtk 
can also be used instead of gnunet-download.
If youever have to abort a download, you can at any time continue it by re\-issuing gnunet\-download with the same filename. In that case GNUnet will not download blocks again that are already present. GNUnets file\-encoding will ensure file integrity, even if the existing file was not downloaded from GNUnet in the first place. Temporary information will be stored in FILENAME.X files until the download is completed. These files are used only if a download is resumed later. If you abort a download for good, you should remember to delete these files.

.SH SETTING ANONYMITY LEVEL

The \fB\-a\fR option can be used to specify additional anonymity
constraints. If set to 0, GNUnet will try to download the
file as fast as possible without any additional
slowdown by the anonymity code. Note that you
will still have a fair degree of anonymity depending
on the current network load and the power of the
adversary. The download is still unlikely to be
terribly fast since the sender may have requested
sender\-anonymity and since in addition to that,
GNUnet will still do the anonymous routing.

This option can be used to limit requests further
than that. In particular, you can require GNUnet to
receive certain amounts of traffic from other peers
before sending your queries. This way, you can gain
very high levels of anonymity \- at the expense of
much more traffic and much higher latency. So set
it only if you really believe you need it.

The definition of ANONYMITY\-RECEIVE is the following:
  If the value v is < 1000, it means that if GNUnet
  routes n bytes of messages from foreign peers, it
  may originate n/v bytes of queries in the same time\-period.
  The time\-period is twice the average delay that GNUnet
  deferrs forwarded queries.

  If the value v is >= 1000, it means that if GNUnet
  routes n bytes of QUERIES from at least (v % 1000)
  peers, it may originate n/v/1000 bytes of queries
  in the same time\-period.

The default is 0 and this should be fine for most
users. Also notice that if you choose values above
1000, you may end up having no throughput at all,
especially if many of your fellow GNUnet\-peers do
the same.

.SH FILES
.TP
~/.gnunet/gnunet.conf
GNUnet configuration file
.SH "REPORTING BUGS"
Report bugs by using mantis <http://www.ovmj.org/~mantis/> or by sending electronic mail to <gnunet\-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet\-gtk\fP(1), \fBgnunet\-insert\fP(1), \fBgnunet\-gtk\fP(1), \fBgnunet\-search\fP(1), \fBgnunet\-download\fP(1), \fBgnunet.conf\fP(5), \fBgnunetd\fP(1)


ChangeLog
Mon Feb 28 00:33:03 EST 2005
	Releasing GNUnet 0.6.6b.

Sun Feb 27 19:27:44 EST 2005
	Fixing mantis 839.

Sun Feb 27 18:44:11 EST 2005
	Fixing mantis 833/834.

Sun Feb 27 18:15:02 EST 2005
	Fixing serious problem with LE in gnunet-insert.
	Fixing mantis 828.

Thu Feb 17 01:14:30 EST 2005
	Various memory leaks for Win32 fixed.
	Added support for GNUNET(D)_CONFIG environment variable.

Wed Feb  9 13:33:38 EST 2005
        Fixed problems with the logrotate code.  Releasing GNUnet 0.6.6a.

Sun Jan 11 19:58:13 CET 2005
	Releasing GNUnet 0.6.6

Sun Jan 11 19:24:27 CET 2005
        gnunet-gtk now resumes aborted downloads automatically.

Sun Nov 14 16:49:02 EST 2004
	Debugged collections, seems to work (not that it looks pretty,
	but it basically works, the rest is UI work).  Fixed deadlock
	in gnunet-gtk logging.  Releasing GNUnet 0.6.5.

Thu Nov  4 09:04:48 EST 2004
	Added (highly experimental) support for collections.

Sat Sep 25 22:29:48 EST 2004
	Releasing GNUnet 0.6.4a.

Sun Sep 19 23:39:04 EST 2004
	Various code cleanups and minor bugfixes.

Sun Aug 29 15:11:28 IST 2004
	Releasing GNUnet 0.6.4.

Tue Aug 24 20:59:14 IST 2004
	Significant enhancements to namespaces.  A namespace can
	now be annotated with meta-data at the time that it is
	created.
	
Sat Aug 21 01:00:06 IST 2004
	Adding GNU gettext support to configure.
	
Thu Aug 19 01:26:55 IST 2004
	Migrating gnunet-gtk to GTK 2.0 (tested with 2.4).

Mon Aug 16 00:30:59 IST 2004
	Removed gcry.  Added dependency on GNU MP library (libgmp)
	for kblocks.

Tue Aug  3 17:57:43 IST 2004
	Fixed bugs in gnunet-update.  Fixed segfault in AFS startup.
	Fixed warnings in gnunet-peer-info/gnunetd.  Fixed packaging
	error (not all of util/gcry included in distribution).
	Releasing GNUnet 0.6.3a.

Mon Aug  2 20:06:07 IST 2004
	Releasing GNUnet 0.6.3.

Sun Aug  1 01:31:19 IST 2004
	Changed most places from HexName to the new, shorter EncName.
	Pushed URIs for namespaces and keyword searches into more
	widespread use.  Replaced all occurences of sprintf with
	SNPRINTF (to obtain extra bounds check).
	
Sun Jul 11 13:54:01 IST 2004
	Moved testcases for libgnunet_util to util/ directory.
	Made DB testcases in afs/module/ run for each available
	database and not just the one configured in /etc/gnunet.conf.

Fri Jun 25 21:05:59 IST 2004
	Releasing GNUnet 0.6.2b (resolves various minor bugs).

Wed May  5 17:34:39 EST 2004
	Releasing GNUnet 0.6.2a (resolves libgcrypt and packageing
	issues, improved gtk-statistics, other minor bugfixes).
	
Fri Apr 30 21:36:23 EST 2004
	Fixed synchronization problem in cron (only relevant for
	parallel downloading).  Releasing GNUnet 0.6.2.

Mon Apr 26 21:13:45 EST 2004
	Added support to gnunet-check to convert file index database
	to new design.  Removed support for pre 0.6.1b bloomfilter
	conversion (would have been too messy to support both).
	Added old-version check to gnunetd startup to make updates
	more systematic in the future (not perfect, though).

Mon Apr 26 19:27:29 EST 2004
	Recursive insertion of directories with gnunet-gtk 
	now works including the progress bar.

Sun Apr 25 18:24:55 EST 2004
	Global changes to fix bug #698.  The fix is still 
	incomplete with respect to pthread.h specific structs.
	
Sun Apr 25 15:20:29 EST 2004
	Fixed compile error in IPv6 transports.

Fri Apr 23 23:38:01 EST 2004
	Added recursive insertion of directories to gnunet-gtk.
	Progress-bar does not work yet, also libextractor is
	always used (no option to turn it off).
	
Fri Apr 23 21:49:18 EST 2004
	Removed need to specify _protocol in configuration file
	(shorter is better).

Wed Apr 21 23:32:36 EST 2004
	Added code to bound priority used by clients by twice the
	average priority of requests from other peers that are 
	routed at the moment.  That prevents clients from exposing
	themselves with excessively high priorities (and also 
	avoids giving too much credit/trust needlessly).

Tue Apr 20 22:27:19 EST 2004
	Added possibility of using a link instead of making a copy
	if gnunetd and the insert-client both run on the same machine.

Tue Apr 20 00:00:58 EST 2004
	Added network transparency for indexing files (by copying
	them to a GNUnet(d) specific directory).  This breaks
	unindexing (for files indexed with previous versions) and
	forces the user to make a copy of the file.  In the near
	future we should resolve this problem by allowing the use
	of a link.

Sun Apr 18 17:24:34 EST 2004
	Adding recursive, parallel download of directories to
	gnunet-download

Sun Apr 18 01:57:32 EST 2004
	gnunet-transport-check can now be used to check
	external connectivity (with other peers).
	
Sat Apr 17 17:46:19 EST 2004
	Reduced memory consuption for AFS insertion and
	download.
	
Sat Apr 10 21:17:46 EST 2004
	Implemented statistics-plot in gnunet-gtk based on
	gnome-system-monitor code.

Fri Apr  9 22:01:51 EST 2004
	Added support for libextractor > 0.2.7.
	
Fri Apr  9  0:29:20 EST 2004
	Completed --help conversion.

Wed Mar 31 15:40:51 EST 2004
	Releasing 0.6.1d.

Tue Mar 30 22:42:13 EST 2004
	Integrated gnunet-pseudonym-create, -delete and -list into
	gnunet-pseudonym as well as gnunet-directory-list, -emptydb
	and -print into gnunet-directory.

Tue Mar  2 10:46:01 EST 2004
	Fixing bug preventing download for indexed files <= 1k.

Sat Feb 28 03:43:34 EST 2004
	Releasing 0.6.1c.

Sat Feb 21 06:16:47 EST 2004
	Modified requestmanager to improve performance at the end of the 
	download.

Thu Feb 19 00:48:33 EST 2004
	Fixed integer underflow possibly causing slow download performance.
	
Sun Feb 15 15:17:05 EST 2004
	Fixed weak key failures with libgcrypt.

Sun Feb 15 13:14:44 EST 2004
	New routing code (untested) commited.

Sun Feb 15 11:58:10 EST 2004
	Fixing problem with 64-bit size_t on alpha (#655).

Sat Jan 31 04:32:33 EST 2004
	Releasing 0.6.1b.

Fri Jan 23 00:00:09 EST 2004
	Fixed indexed content being overwritten by migrated content
	
Thu Jan 22 19:27:33 EST 2004
	Fixed integer-overflow problems in statuscalls.

src/applications/afs/tools/gnunet-download.c
/*
     This file is part of GNUnet.
     (C) 2001, 2002, 2003, 2004 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file applications/afs/tools/gnunet-download.c 
 * @brief Main function to download files from GNUnet.
 * @author Christian Grothoff
 **/

#include "gnunet_afs_esed2.h"
#include "platform.h"

/**
 * Prints the usage information for this command if the user errs.
 * Aborts the program.
 **/
static void printhelp() {
  static Help help[] = {
    { 'a', "anonymity", "LEVEL",
      "set the desired LEVEL of receiver-anonymity" },
    HELP_CONFIG,
    HELP_HELP,
    HELP_HOSTNAME,
    HELP_LOGLEVEL,
    { 'o', "output", "FILENAME",
      "write the file to FILENAME (required)" },
    { 'R', "recursive", NULL,
      "download a GNUnet directory recursively" },
    { 't', "threads", "NUMBER",
      "specifies the NUMBER of files that maybe downloaded in parallel for a recursive download"},
    HELP_VERSION,
    HELP_VERBOSE,
    HELP_END,
  };
  formatHelp("gnunet-download [OPTIONS] GNUNET-URI",
	     "Download file from GNUnet.",
	     help);
}

/**
 * ParseOptions for gnunet-download.
 * @return SYSERR to abort afterwards, OK to continue
 **/
static int parseOptions(int argc,
			char ** argv) {
  int c;  

  while (1) {
    int option_index = 0;
    static struct GNoption long_options[] = {
      LONG_DEFAULT_OPTIONS,
      { "anonymity", 1, 0, 'a' }, 
      { "output",    1, 0, 'o' },
      { "recursive", 0, 0, 'R' },
      { "threads",   1, 0, 't' },
      { "verbose",   0, 0, 'V' },
      { 0,0,0,0 }
    };    
    c = GNgetopt_long(argc,
		      argv, 
		      "a:vhdc:L:H:Vo:Rt:", 
		      long_options, 
		      &option_index);    
    if (c == -1) 
      break;  /* No more flags to process */
    if (YES == parseDefaultOptions(c, GNoptarg))
      continue;
    switch(c) {
    case 'a': {
      unsigned int receivePolicy;
      
      if (1 != sscanf(GNoptarg, "%ud", &receivePolicy)) {
        LOG(LOG_FAILURE,
	    "FAILURE: You must pass a number to the -a option.\n");
	return -1;
      }
      setConfigurationInt("AFS",
      			  "ANONYMITY-RECEIVE",
			  receivePolicy);
      break;
    }
    case 't': {
      unsigned int threads;
      
      if (1 != sscanf(GNoptarg, "%ud", &threads)) {
        LOG(LOG_FAILURE,
	    "FAILURE: You must pass a number to the -t option.\n");
	return -1;
      }
      if (threads == 0) {
	threads = 1; /* actual minimum value */
      }
      setConfigurationInt("GNUNET-DOWNLOAD",
      			  "PARALLELIZATION",
			  threads);
      break;
    }
    case 'R': 
      FREENONNULL(setConfigurationString("GNUNET-DOWNLOAD",
					 "RECURSIVE",
					 "YES"));
      break;
    case 'o':
      FREENONNULL(setConfigurationString("GNUNET-DOWNLOAD",
					 "FILENAME",
					 GNoptarg));
      break;
    case 'v': 
      printf("GNUnet v%s, gnunet-download v%s\n",
	     VERSION,
	     AFS_VERSION);
      return SYSERR;
    case 'V':
      FREENONNULL(setConfigurationString("GNUNET-DOWNLOAD",
					 "VERBOSE",
					 "YES"));
      break;
    case 'h': 
      printhelp(); 
      return SYSERR;
    default: 
      LOG(LOG_FAILURE,
	  "Unknown option %c. Aborting.\n"\
	  "Use --help to get a list of options.\n",
	  c);
      return SYSERR;
    } /* end of parsing commandline */
  } /* while (1) */
  if (argc - GNoptind != 1) {
    LOG(LOG_WARNING, 
	"Not enough arguments. "\
	"You must specify atleast a GNUnet AFS URI\n");
    printhelp();
    return SYSERR;
  }
  FREENONNULL(setConfigurationString("GNUNET-DOWNLOAD",
				     "URI",
				     argv[GNoptind++]));
  return OK;
}

typedef struct {
  Semaphore * sem;
  char * filename;
  FileIdentifier * fid;
  cron_t startTime;
  size_t lastProgress;
  int result;
  PTHREAD_T thread;
} DownloadInfo;

/**
 * This method is called whenever data is received.
 * The current incarnation just ensures that the main
 * method exits once the download is complete.
 **/
static void progressModel(ProgressStats * stats,
			  DownloadInfo * data) {
  if (YES == testConfigurationString("GNUNET-DOWNLOAD",
				     "VERBOSE",
				     "YES")) {
    if (stats->progress != data->lastProgress)
      printf("Download at %8u out of %8u bytes (%8.3f kbps)\r",
  	     (unsigned int) stats->progress, 
	     (unsigned int) stats->filesize,
	     (stats->progress/1024.0) / 
	     (((double)(cronTime(NULL)-(data->startTime-1))) / (double)cronSECONDS) );
  }
  data->lastProgress = stats->progress;
  if (stats->progress == stats->filesize)
    SEMAPHORE_UP(data->sem);
}

static void scheduleDownload(FileIdentifier * fid,
			     char * filename);

static int downloadFileHelper(DownloadInfo * di) {
  RequestManager * rm;
  int ok;

  if (di->sem != NULL)
    errexit("FATAL: assertion failed!\n");
  cronTime(&di->startTime);
  di->lastProgress = 0;
  di->sem = SEMAPHORE_NEW(0);
  rm = downloadFile(di->fid,
		    di->filename, 
		    (ProgressModel) &progressModel,
		    di);
  if (rm == NULL) {
    printf("Download %s failed (error messages should have been provided).\n",
	   di->filename);
    return SYSERR;
  }
  SEMAPHORE_DOWN(di->sem);
  SEMAPHORE_FREE(di->sem);
  di->sem = NULL;
  destroyRequestManager(rm);
  printf("\nDownload %s %s.  Speed was %8.3f kilobyte per second.\n",
	 di->filename,
	 (ntohl(di->fid->file_length) == di->lastProgress) ?
	 "complete" : "incomplete",
	 (di->lastProgress/1024.0) / 
	 (((double)(cronTime(NULL)-di->startTime)) / (double)cronSECONDS) );
  if (ntohl(di->fid->file_length) == di->lastProgress)
    ok = OK;
  else
    ok = SYSERR;
  if ( (ok == OK) && 
       testConfigurationString("GNUNET-DOWNLOAD",
			       "RECURSIVE",
			       "YES")) {
    /* download files in directory! */
    char * exp;
    GNUnetDirectory * dir;
    unsigned int i;

    exp = expandFileName(di->filename);
    dir = readGNUnetDirectory(exp);
    FREE(exp);
    if (dir == NULL) 
      return ok; /* not a directory */    
    for (i=0;i<ntohl(dir->number_of_files);i++) {
      RootNode * rn;
      char * fn;
      char * rfn;

      rn = &((GNUnetDirectory_GENERIC*)dir)->contents[i];
      rfn = getFilenameFromNode(rn);
      fn = MALLOC(strlen(di->filename) + 
		  strlen(rfn) +
		  128);
      strcpy(fn, di->filename);
      /* remove '.gnd' or add ".dir" */
      if ( (strlen(fn) > 1+strlen(GNUNET_DIRECTORY_EXT)) &&
	   (0 == strcmp(GNUNET_DIRECTORY_EXT, 
			&fn[strlen(fn)-4])) ) {
	fn[strlen(fn)-4] = '\0';
      } else {
	strcat(fn, ".dir");
      }
      mkdirp(fn); /* create directory */
      strcat(fn, "/");
      strcat(fn, rfn);
      FREE(rfn);
      scheduleDownload(&rn->header.fileIdentifier,
		       fn);
      FREE(fn);
    }    
    FREE(dir);
  }
  return ok;
}			 

static DownloadInfo ** pending = NULL;
static unsigned int pendingCount = 0;
static Mutex lock;
/* just not SYSERR/OK */
#define PENDING 42
#define RUNNING 43
#define JOINED 44

static void scheduleDownload(FileIdentifier * fid,
			     char * filename) {
  MUTEX_LOCK(&lock);
  GROW(pending,
       pendingCount,
       pendingCount+1);
  pending[pendingCount-1] = MALLOC(sizeof(DownloadInfo));
  pending[pendingCount-1]->fid = MALLOC(sizeof(FileIdentifier));
  memcpy(pending[pendingCount-1]->fid,
	 fid,
	 sizeof(FileIdentifier));
  pending[pendingCount-1]->filename = STRDUP(filename);  
  pending[pendingCount-1]->result = PENDING;
  pending[pendingCount-1]->sem = NULL;
  MUTEX_UNLOCK(&lock);
}

static Semaphore * semSignal;

static void * process(DownloadInfo * di) {
  di->result = downloadFileHelper(di);
  SEMAPHORE_UP(semSignal);
  return NULL;
}

static int run(int threadLimit) {
  unsigned int i;
  unsigned int left;
  unsigned int running;
  int ret;    
  void * unused;

  if (threadLimit == 0) /* should never happen */
    threadLimit = 1; /* need at least 1! */

  ret = OK;
  semSignal = SEMAPHORE_NEW(threadLimit);
  left = pendingCount;
  running = 0;
  while ( (left > 0) || (running > 0) ) {
    if (left > 0) {
      SEMAPHORE_DOWN(semSignal);
      for (i=0;i<pendingCount;i++) {
	if (pending[i]->result == PENDING) {
	  pending[i]->result = RUNNING;
	  if (0 != PTHREAD_CREATE(&pending[i]->thread,
				  (PThreadMain) &process,
				  pending[i],
				  16*1024))
	    errexit("FATAL: could not create download thread: %s\n",
		    strerror(errno));
	  break;
	}
      }
    }
    left = 0;
    running = 0;
    for (i=0;i<pendingCount;i++) {
      if (pending[i]->result == PENDING) {
	left++;
	continue;
      }
      if (pending[i]->result == RUNNING) {
	running++;
	continue;
      }
      if (pending[i]->result == JOINED)
	continue;
      if (pending[i]->result == SYSERR)
	ret = SYSERR;
      PTHREAD_JOIN(&pending[i]->thread,
		   &unused);
      pending[i]->result = JOINED;
      FREE(pending[i]->fid);
      FREE(pending[i]->filename);
    }
    /* wait a bit */
    if ( (left == 0) && (running > 0) ) 
      gnunet_util_sleep(150 * cronMILLIS);        
  }  
  /* wait for all threads to terminate */
  for (i=0;i<threadLimit;i++)
    SEMAPHORE_DOWN(semSignal); 
  /* join on all */
  for (i=0;i<pendingCount;i++) {
    if (pending[i]->result == PENDING)		
      errexit("FATAL: assertion failed at %s:%d\n",
	      __FILE__, __LINE__);
    if (pending[i]->result == JOINED)
      continue;
    if (pending[i]->result == SYSERR)
      ret = SYSERR;    
    FREE(pending[i]->fid);
    FREE(pending[i]->filename);
    PTHREAD_JOIN(&pending[i]->thread,
		 &unused);
    pending[i]->result = JOINED;
  }
  for (i=0;i<pendingCount;i++) 
    FREE(pending[i]);  
  GROW(pending,
       pendingCount,
       0);
  SEMAPHORE_FREE(semSignal);
  return ret;
}

/**
 * Main function to download files from GNUnet.
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return return value from download file: 0: ok, -1, 1: error
 **/   
int main(int argc, 
	 char ** argv) {
  char * fstring;
  FileIdentifier * fid;
  char * filename;
  int ok;
  int threadLimit;

  if (SYSERR == initUtil(argc, argv, &parseOptions))
    return 0;

  threadLimit = getConfigurationInt("GNUNET-DOWNLOAD",
				    "PARALLELIZATION");
  if (threadLimit == 0)
    threadLimit = 30; /* default */
  filename = getConfigurationString("GNUNET-DOWNLOAD",
				    "FILENAME");
  if (filename == NULL) {
    LOG(LOG_ERROR,
	"ERROR: You must specify a filename (option -o).\n");
    printhelp();
    return -1;
  }
  fstring = getConfigurationString("GNUNET-DOWNLOAD",
  				   "URI");
  fid = stringToFileIdentifier(fstring);
  FREE(fstring);
  if (fid == NULL) {
    LOG(LOG_ERROR,
        "ERROR: Can't proceed without valid URI.\n");
    FREE(filename);
    return -1;
  }

  MUTEX_CREATE_RECURSIVE(&lock);
  startAFSPriorityTracker();
  startCron();
  initAnonymityPolicy(NULL);
  scheduleDownload(fid,
		   filename);
  FREE(fid);
  FREE(filename);
  ok = run(threadLimit);
  doneAnonymityPolicy();
  MUTEX_DESTROY(&lock);
  stopCron();
  stopAFSPriorityTracker();
  doneUtil();
  
  if (ok == OK)
    return 0;
  else
    return 1;
}

/* end of gnunet-download.c */
